### PR TITLE
BUG: DataFrame.to_dict when orient=index data loss

### LINKED ISSUE #22801: DataFrame.to_dict can result in data loss when indexes are not unique
#### Code Sample,

```python
df = pd.DataFrame([["A","B","C"],["A","D","E"],["F","G","K"]])
df.set_index(0, inplace=True)
print(df)
```
---1  2
0      
A  B  C
A  D  E
F  G  K
```python
print(df.to_dict(orient="index"))
```
{'A': {1: 'D', 2: 'E'}, 'F': {1: 'G', 2: 'K'}}

#### Problem description

The to_dict method silently tosses away data when indexes are not unique. At least in my impression the rest of the API doesn’t mind non-unique index’s much, so I don’t like this behavior. I think that it should either preserve the data, or raise an an error / warning saying that to_dict requires unique indexes. 

#### Expected Output
'''
{'A': [{1: 'B', 2: 'C'},{1: 'D', 2: 'E'}], 'F': {1: 'G', 2: 'K'}}

or at least throw an error saying ".to_dict(orient="index") does not support non-unique indexes" 

'''
#### Output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.5.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.23.4
pytest: None
pip: 10.0.1
setuptools: 40.0.0
Cython: 0.28.5
numpy: 1.15.0
scipy: None
pyarrow: 0.9.0
xarray: None
IPython: 6.5.0
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by jameswinegar**
- [x] closes #22801 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -1163,8 +1163,13 @@ def to_dict(self, orient='dict', into=dict):
                            for k, v in zip(self.columns, np.atleast_1d(row)))
                     for row in self.values]
         elif orient.lower().startswith('i'):
-            return into_c((t[0], dict(zip(self.columns, t[1:])))
-                          for t in self.itertuples())
+            if self.index.unique().size == self.index.size:
```
Comment (jreback) [inline pandas/core/frame.py]: you can just use ``self.index.is_unique``

File: pandas/core/frame.py [hunk 2]
```diff
@@ -1163,8 +1163,13 @@ def to_dict(self, orient='dict', into=dict):
...
+                return into_c((t[0], dict(zip(self.columns, t[1:])))
+                              for t in self.itertuples())
+            else:
```
Comment (jreback) [inline pandas/core/frame.py]: no need for an else!

File: pandas/tests/frame/test_convert_to.py [hunk 3]
```diff
@@ -71,6 +71,10 @@ def test_to_dict_timestamp(self):
         tm.assert_dict_equal(test_data_mixed.to_dict(orient='split'),
                              expected_split_mixed)
 
+    def test_to_dict_index_not_unique_with_index_orient(self):
+        df = DataFrame({'a': [1, 2], 'b': [0.5, 0.75]}, index=['A', 'A'])
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: can you add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a note in the api changes section about this for 0.24.0

**Review by jreback — CHANGES_REQUESTED** [review]
does this affect any other orient types?

can you add some equiv tests for .to_json? (or is this affected?)

**Comment by jameswinegar** [comment]
@jreback on .to_json https://github.com/pandas-dev/pandas/pull/4376

**Comment by jameswinegar** [comment]
@jreback Where are the API changes documentation?

File: doc/source/whatsnew/v0.24.0.txt [hunk 4]
```diff
@@ -191,6 +191,7 @@ Other Enhancements
 - :meth:`Series.resample` and :meth:`DataFrame.resample` have gained the :meth:`Resampler.quantile` (:issue:`15023`).
 - :meth:`Index.to_frame` now supports overriding column name(s) (:issue:`22580`).
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
+- :func:`~DataFrame.to_dict` with orient on index and a non-unique index will raise a ValueError instead of losing data (:issue:`22801`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: move to other api changes

**Comment by jameswinegar** [comment]
@jreback any more edits needed?

File: doc/source/whatsnew/v0.24.0.txt [hunk 5]
```diff
@@ -804,6 +804,7 @@ Reshaping
 - Bug in :meth:`Series.replace` and meth:`DataFrame.replace` when dict is used as the ``to_replace`` value and one key in the dict is is another key's value, the results were inconsistent between using integer key and using string key (:issue:`20656`)
 - Bug in :meth:`DataFrame.drop_duplicates` for empty ``DataFrame`` which incorrectly raises an error (:issue:`20516`)
 - Bug in :func:`pandas.wide_to_long` when a string is passed to the stubnames argument and a column name is a substring of that stubname (:issue:`22468`)
+- Bug in :func:`DataFrame.to_dict` raises ``ValueError`` when used with ``orient='index'`` and a non-unique index instead of losing data (:issue:`22801`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: move to API Breaking changes

File: pandas/tests/frame/test_convert_to.py [hunk 6]
```diff
@@ -71,6 +71,12 @@ def test_to_dict_timestamp(self):
         tm.assert_dict_equal(test_data_mixed.to_dict(orient='split'),
                              expected_split_mixed)
 
+    def test_to_dict_index_not_unique_with_index_orient(self):
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: is this raise only happening on orient='index' or other orients as well?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_convert_to.py (hunk 6 above)
Comment (jameswinegar) [reply pandas/tests/frame/test_convert_to.py]: Yes

**Comment by jameswinegar** [comment]
@jreback is that move good?

File: doc/source/whatsnew/v0.24.0.txt [hunk 7]
```diff
@@ -368,6 +368,37 @@ is the case with :attr:`Period.end_time`, for example
 
    p.end_time
 
+.. _whatsnew_0240.api_breaking.frame_to_dict_index_orient:
+
+Raise ValueError in ``DataFrame.to_dict(orient='index')``
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.txt]: A dedicated section like this is typically reserved for larger changes, which I don't think applies here. Can you instead just add this as a one-line note in `Other API Changes`?

File: pandas/tests/frame/test_convert_to.py (hunk 6 above)
Comment (WillAyd) [reply pandas/tests/frame/test_convert_to.py]: Can you parametrize this function to test all of the orients? FWIW this probably should be allowed with `orient="table"` so curious how that behaves

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 7 above)
Comment (WillAyd) [reply doc/source/whatsnew/v0.24.0.txt]: My mistake I didn't see that comment; with that being the case go ahead and leave as is

File: pandas/tests/frame/test_convert_to.py (hunk 6 above)
Comment (WillAyd) [reply pandas/tests/frame/test_convert_to.py]: Hmm nevermind - I suppose this change is only applicable to `index` orients given that's the only one with the index value as a key of the dict

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

File: pandas/tests/frame/test_convert_to.py (hunk 3 above)
Comment (jameswinegar) [reply pandas/tests/frame/test_convert_to.py]: Added.

File: doc/source/whatsnew/v0.24.0.txt [hunk 8]
```diff
@@ -847,7 +847,7 @@ Reshaping
 - Bug in :meth:`Series.replace` and meth:`DataFrame.replace` when dict is used as the ``to_replace`` value and one key in the dict is is another key's value, the results were inconsistent between using integer key and using string key (:issue:`20656`)
 - Bug in :meth:`DataFrame.drop_duplicates` for empty ``DataFrame`` which incorrectly raises an error (:issue:`20516`)
 - Bug in :func:`pandas.wide_to_long` when a string is passed to the stubnames argument and a column name is a substring of that stubname (:issue:`22468`)
--
+- Bug in :func:`merge` when merging ``datetime64[ns, tz]`` data that contained a DST transition (:issue:`18885`)
```
Comment (jameswinegar) [inline doc/source/whatsnew/v0.24.0.txt]: Resolved merge conflict

File: doc/source/whatsnew/v0.24.0.txt [hunk 9]
```diff
@@ -372,6 +372,37 @@ is the case with :attr:`Period.end_time`, for example
...
+.. _whatsnew_0240.api_breaking.frame_to_dict_index_orient:
+
+Raise ValueError in ``DataFrame.to_dict(orient='index')``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Bug in :func:`DataFrame.to_dict` raises ``ValueError`` when used with 
+``orient='index'`` and a non-unique index instead of losing data (:issue:`22801`)
+
+Previous Behavior:
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: i don't think we need this whole section, just a single line entry in API breaking changes. sorry about that.

**Review by jreback — CHANGES_REQUESTED** [review]
rebase and ping on green.

**Comment by jameswinegar** [comment]
@jreback fixed. long wait last night for circleci

File: doc/source/whatsnew/v0.24.0.txt [hunk 10]
```diff
@@ -373,6 +373,14 @@ is the case with :attr:`Period.end_time`, for example
...
    p.end_time
 
+.. _whatsnew_0240.api_breaking.frame_to_dict_index_orient:
+
+Raise ValueError in ``DataFrame.to_dict(orient='index')``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Bug in :func:`DataFrame.to_dict` raises ``ValueError`` when used with 
+``orient='index'`` and a non-unique index instead of losing data (:issue:`22801`)
+
```
[the remark this answers is not in the document]
Comment (jameswinegar) [reply doc/source/whatsnew/v0.24.0.txt]: Added an ipython block that will raise the error. @jreback

[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: right but now the docs will not work. so, you can do this as an ipython block but you need an ``:okexcept:`` to handle the exception.

also show the df after defining it

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt [hunk 11]
```diff
@@ -373,6 +373,22 @@ is the case with :attr:`Period.end_time`, for example
...
+.. ipython:: python
+:okexcept:
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: line this up as below, e.g.
```
.. ipython:: python
    :okexcept:

    df = ......
```

**Review by jreback — APPROVED** [review]
one more doc correction, ping on green.

**Comment by jreback** [comment]
thanks @jameswinegar

**Pull request merged**